set inventory on slurmd relation

jamesbeedy opened the pull request:
set inventory data on each individual unit on the slurmd relation.

* remove unused utility functions
* remove unused slurmd_peer relation code
* alter the way in which inventory data is assembled in slurmctld

jamesbeedy committed:
These changes remove the slurmd peer relation and in its place
set inventory data on each individual unit on the slurmd relation.

* remove unused utility functions
* remove slurmd_peer relation code and interface from metadata.yaml
* alter the way in which inventory data is assembled in slurmctld

heitorPB commented on charm-slurmd/src/interface_slurmd.py:
```diff
@@ -158,12 +191,28 @@ def _store_slurmctld_host_port(self, host, port):
+        """Retrieve the munge_key from the StoredState."""
+        return self._stored.munge_key
+
+    def get_node_inventory(self):
+        """Retrieve the node inventory from the relation."""
+        return self._relation.data[self.model.unit].get("inventory")
```
Do we need this method? We already have a `node_inventory` property.

jamesbeedy replied:
good point

heitorPB commented:
Tests pass on focal. Will run on centos later.

heitorPB commented on charm-slurmd/src/interface_slurmd.py:
```diff
@@ -158,12 +183,19 @@ def _store_slurmctld_host_port(self, host, port):
+        """Set this node as not new and trigger a reconfiguration."""
+        inv = self.node_inventory
+        inv["new_node"] = False
+        self._relation.data[self.model.unit]['inventory'] = json.dumps(inv)
```
```suggestion
        self.node_inventory = inv
```

We have the property (setter and getter) for `node_inventory`, is there a reason to not use here?

heitorPB commented on charm-slurmd/src/interface_slurmd.py:
```diff
@@ -132,6 +137,26 @@ def slurmctld_port(self):
         return self._stored.slurmctld_port
 
+    @property
+    def node_inventory(self):
```
```suggestion
    def node_inventory(self) -> dict:
```

I think it makes development easier if we start adding type hints. What do you think about it?

heitorPB commented on charm-slurmd/src/interface_slurmd.py:
```diff
@@ -132,6 +137,26 @@ def slurmctld_port(self):
+        self._relation.data[self.model.unit]["inventory"] = json.dumps(inventory)
+
+    @property
+    def partition_name(self):
```
```suggestion
    def partition_name(self) -> str:
```

heitorPB reviewed (changes requested):
Crashing on centos:

```bash
Traceback (most recent call last):
  File "./src/charm.py", line 317, in <module>
    main(SlurmdCharm)
  File "/var/lib/juju/agents/unit-slurmd-25/charm/venv/ops/main.py", line 406, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-slurmd-25/charm/venv/ops/main.py", line 140, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-slurmd-25/charm/venv/ops/framework.py", line 278, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-slurmd-25/charm/venv/ops/framework.py", line 722, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-slurmd-25/charm/venv/ops/framework.py", line 767, in _reemit
    custom_handler(event)
  File "/var/lib/juju/agents/unit-slurmd-25/charm/src/interface_slurmd.py", line 102, in _on_relation_joined
    self._store_munge_key(app_data["munge_key"])
TypeError: string indices must be integers
```

heitorPB commented on charm-slurmd/src/interface_slurmd.py:
```diff
@@ -60,56 +63,58 @@ def __init__(self, charm, relation_name):
+        # at this point so retrieve them from the relation data and store
+        # them in the charm's stored state.
+        self._store_munge_key(app_data["munge_key"])
+        self._store_slurmctld_host_port(
+            app_data["slurmctld_host"], app_data["slurmctld_port"]
+        )
```
What do you think about:

```suggestion
        self._store_slurmctld_host_port(app_data["slurmctld_host"], 
                                        app_data["slurmctld_port"])

```

jamesbeedy committed:
Update charm-slurmd/src/interface_slurmd.py

Co-authored-by: Heitor <heitorpbittencourt@gmail.com>

jamesbeedy committed:
Update charm-slurmd/src/interface_slurmd.py

Co-authored-by: Heitor <heitorpbittencourt@gmail.com>

jamesbeedy committed:
Update charm-slurmd/src/interface_slurmd.py

Co-authored-by: Heitor <heitorpbittencourt@gmail.com>

jamesbeedy committed:
Update charm-slurmd/src/interface_slurmd.py

Co-authored-by: Heitor <heitorpbittencourt@gmail.com>

jamesbeedy committed:
remove typo and add type annotations

heitorPB commented on charm-slurmd/src/interface_slurmd.py:
```diff
@@ -113,27 +113,27 @@ def _on_relation_broken(self, event):
 
     @property
-    def _relation(self):
+    def _relation(self) -> Relation:
```
:+1:

heitorPB commented:
tests pass on CentOS and removing the leader does not crash jobs :tada: 

@jamesbeedy  could you lint the charms and an entry to the CHANGELOG? Something in the lines of _fix removal of slurmd/leader without crashing running jobs_.

jamesbeedy committed:
update changelog

heitorPB commented on CHANGELOG:
```diff
@@ -7,6 +7,13 @@ This file keeps track of all notable changes to the Slurm Charms.
 Unreleased
 ----------
 
+- fix breakage in charms when the slurmd leader is removed
+- remove slurmd peer relation in favor of the slurmd relation
+- slurmd units set their own inventory on the slurmd relation unit data 
+
+Unreleased
+----------
+
```
```suggestion
- fix breakage in charms when the slurmd leader is removed
- remove slurmd peer relation in favor of the slurmd relation
```

We don't need to repeat the `Unreleased` header (and i guess it would break our release script).

Also I don't think it matters for end users where internal data lives. The CHANGELOG is copied to the osd-docs, to inform end users, not charm developers, of notable changes in the user-facing parts of project.

This should be written somewhere, i will start a "contributing" page with this info.

jamesbeedy committed:
Update CHANGELOG

Co-authored-by: Heitor <heitorpbittencourt@gmail.com>

heitorPB reviewed (approved):
That's a huge improvement! Nice work!

And tests pass on focal.

heitorPB commented:
```
machine-96: 15:27:42 ERROR unit.slurmd/30.juju-log Uncaught exception while in charm code:
Traceback (most recent call last):
  File "./src/charm.py", line 317, in <module>
    main(SlurmdCharm)
  File "/var/lib/juju/agents/unit-slurmd-30/charm/venv/ops/main.py", line 406, in main
    _emit_charm_event(charm, dispatcher.event_name)
  File "/var/lib/juju/agents/unit-slurmd-30/charm/venv/ops/main.py", line 140, in _emit_charm_event
    event_to_emit.emit(*args, **kwargs)
  File "/var/lib/juju/agents/unit-slurmd-30/charm/venv/ops/framework.py", line 278, in emit
    framework._emit(event)
  File "/var/lib/juju/agents/unit-slurmd-30/charm/venv/ops/framework.py", line 722, in _emit
    self._reemit(event_path)
  File "/var/lib/juju/agents/unit-slurmd-30/charm/venv/ops/framework.py", line 767, in _reemit
    custom_handler(event)
  File "./src/charm.py", line 159, in _on_config_changed
    self._get_set_partition_name()
  File "./src/charm.py", line 299, in _get_set_partition_name
    if not self.get_partition_name():
  File "./src/charm.py", line 304, in get_partition_name
    return self._slurmd.partition_name
  File "/var/lib/juju/agents/unit-slurmd-30/charm/src/interface_slurmd.py", line 153, in partition_name
    return self._relation.data[self.model.app].get('partition-name')
AttributeError: 'NoneType' object has no attribute 'data'
unit-slurmd-30: 15:27:42 ERROR juju.worker.uniter.operation
```

we need that `if relation`, @jamesbeedy :(

heitorPB committed:
set slurmdbd_available to false when it leaves

jamesbeedy committed:
re-add the peer relation for tracking partition_name

These changes move the storing and accessing of the partition_name
to the peer relation application data. We use this convention
such that we can persist the partition_name across leadership
changes and as a place to store and update the partition_name
in a central location.

jamesbeedy committed:
update changelog

jamesbeedy committed:
check for the munge-key before we try and get and sleep more

Add additional check for the munge_key in the application data
before tyrying to access it. Also, sleep for longer and add more
retries to try and start slurmd.

jamesbeedy committed:
don't assume the inventory always exists

heitorPB reviewed (approved):
Tests pass locally on CentOS7 and Ubuntu Focal :tada: 

Nice work, @jamesbeedy !
